_posts/2024-11-22-cheri-myths-safety-critical.markdown

```diff
@@ -0,0 +1,183 @@
+
+In the fourth case, CHERI improves things because now it should show up in testing reliably and you can fix the issue.
+
+The baseline is not better in *any* of these cases.
+The example that people often use here is anti-lock braking systems (ABS).
+You absolutely do not want the controller for ABS to crash.
+But will it if you run it on a CHERI system?
+
+If it's written in MISRA C and passing the checks (and not arbitrarily silencing them), it won't.
+If it's written in safe Rust, it won't.
+If it's written in SPARK Ada, it won't.
+
+Something like this will not be dynamically allocating memory, it will not be accessing arbitrary memory locations via untrusted pointers.
+It will have a set of inputs that should be amenable to exhaustive testing.
+
+If you have a chip dedicated to running the ABS and that code is written in a safe language, CHERI doesn't improve on the state of the art: you have already statically proven that nothing that CHERI dynamically prevents can happen.
+If you are building a *mixed criticality* system, the benefits of CHERI are far more pronounced: CHERI will allow you to collocate things on the same chip that previously needed to be isolated.
+
+## What does CHERIoT do?
+
+CHERIoT RTOS provides three ways for a compartment to handle CHERI failures (and other synchronous trap, such as invalid instruction errors):
+
+ - If there is enough stack space, a compartment may provide an error handler that runs with a copy of the register file and can inspect and modify the state.
+ - If there is not enough stack space or the first kind of error handler is not provided, the compartment may provide a *stackless* error handler.
+   This is used to implement our structured error handling mechanisms.
+ - If neither is provided, or the error handler returns the force-unwind result, the current compartment call is and control returns to the caller with an error code in the return register.
+
+For stateless compartments, the last is often the right approach: if something goes wrong, let the caller know.
+This is often the best fit approach for safety-critical systems, where things should *never* go wrong and you've got static analysis receipts to demonstrate this.
+
+If you're relying on the notion of a 'benign' error, the first kind of error handler helps.
+You can decode the current instruction and determine whether it's a load or store.
+If it's a store, you can skip it, if it's a load you can also also put a zero in the target register.
+Ideally, you'd also write some telemetry that would let you fix the bug.
+This turns your 'benign' memory-safety bugs into ones that have deterministic behaviour that is guaranteed not to impact the rest of the system.
+
+The middle category is most interesting.
+This form uses is most commonly used with `longjmp`, to jump to the nearest error handler on the stack, which would be in violation of MISRA C 21.4, which says that `setjmp.h` may not be used and the functions in it may not be used.
```

Comment:
```suggestion
This form is most commonly used with `longjmp`, to jump to the nearest error handler on the stack, which would be in violation of MISRA C 21.4, which says that `setjmp.h` may not be used and the functions in it may not be used.
```